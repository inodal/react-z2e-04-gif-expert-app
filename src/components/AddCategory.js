import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        // Avoid default behaviour that will submit and refresh whole view
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(cats => [inputValue.trim(), ...cats]); // Show new first
            setInputValue('');
        }
        
    }

    return (
        <>
        <form onSubmit= { handleSubmit }>
        <p>{ inputValue }</p> 
            <input 
                type ="text"
                value = { inputValue }
                onChange = { handleInputChange }
                
            />
        </form>
            
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

