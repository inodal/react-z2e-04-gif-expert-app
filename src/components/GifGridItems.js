import React from 'react'

export const GifGridItems = ({id, title, url }) => {

    console.log(`${id}:\ntitle: ${title} \nurl: ${url}`);
    return (
        <div className="card">
            <img src={ url } alt={ title } />
            <p>{ title }</p>
        </div>
    )
}
