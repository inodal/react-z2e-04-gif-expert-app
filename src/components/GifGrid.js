import React, { useState, useEffect } from 'react'
import { GifGridItems } from './GifGridItems';

export const GifGrid = ({category}) => {

    
    const [images, setImages] = useState([]);

    // Use hook 'useEffect' to render only the first time
    useEffect(() => {
        getGifs()
    }, [])

    const getGifs = async () => {
        const url = 'https://api.giphy.com/v1/gifs/search?&limit=10&q=Dragon+Ball&api_key=a6moXgoQ6vuFxhVHum2nE6CPtZdV66Uz';
        const resp = await fetch(url);
        const { data } = await resp.json();

        const gifs = data.map( img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }

        });
        console.log(gifs);
        setImages(gifs);
    };

    return (
        <div>
            <h3> { category } </h3>
            {
                images.map(img => (
                    <GifGridItems 
                        key={ img.id }
                        { ...img } 
                    />
                ))
            }
        </div>
    )
}
