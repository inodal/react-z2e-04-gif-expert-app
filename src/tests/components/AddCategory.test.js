import { shallow } from 'enzyme'
import React from 'react'
import { AddCategory } from '../../components/AddCategory'
import '@testing-library/jest-dom';

describe('AddCategory component test suit', () => {
    
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    
    beforeEach( () => {
        jest.clearAllMocks(); 
        // Initialize wrapper, e.g. to clean inputValue after text box test
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);

    })

    test('should show AddCategory component properly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should change text box', () => {
        const input = wrapper.find('input');
        const value ='Hello world!';
        const paragraphPrefix = 'GifExpertApp: ';

        input.simulate('change', {target: {value}});

        expect(wrapper.find('p').text().trim()).toBe(paragraphPrefix+value);
    })
    
    test('should NOT post the form on submit', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect(setCategories).not.toHaveBeenCalled();
    })

    test('should call setCategories and clean text box', () => {
        const value = 'Hello world!';

        wrapper.find('input').simulate('change', {target: {value}});

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenLastCalledWith(expect.any(Function));
        expect(wrapper.find('input').text().trim()).toBe('');
    })
    
    
})
