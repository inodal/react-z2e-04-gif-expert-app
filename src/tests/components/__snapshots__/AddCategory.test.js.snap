// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory component test suit should show AddCategory component properly 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Search funny GIFs!! Type a topic and press enter NOW!!"
      type="text"
      value=""
    />
  </form>
  <p
    className="hidden-elem"
  >
    GifExpertApp: 
  </p>
</Fragment>
`;
